controllers/atlasmigration_controller.go

```diff
@@ -16,39 +16,310 @@ limitations under the License.
 package controllers
 
 import (
+	"bytes"
 	"context"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+	"time"
 
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
 	ctrl "sigs.k8s.io/controller-runtime"
 	"sigs.k8s.io/controller-runtime/pkg/client"
 	"sigs.k8s.io/controller-runtime/pkg/log"
 
 	dbv1alpha1 "github.com/ariga/atlas-operator/api/v1alpha1"
+	"github.com/ariga/atlas-operator/internal/atlas"
 )
 
+// CLI is the interface used to interact with Atlas CLI
+type MigrateCLI interface {
+	Apply(ctx context.Context, data *atlas.ApplyParams) (*atlas.ApplyReport, error)
+}
+
 // AtlasMigrationReconciler reconciles a AtlasMigration object
 type AtlasMigrationReconciler struct {
 	client.Client
+	CLI    MigrateCLI
 	Scheme *runtime.Scheme
 }
 
+// HCLTmplData is the data used to render the HCL template
+// that will be used for Atlas CLI
+type HCLTmplData struct {
+	URL       string
+	Migration struct {
+		Dir string
+	}
+	Cloud struct {
+		URL     string
+		Token   string
+		Project string
+	}
+	Data struct {
+		RemoteDir struct {
+			Name string
+			Tag  string
+		}
+	}
+}
+
 //+kubebuilder:rbac:groups=db.atlasgo.io,resources=atlasmigrations,verbs=get;list;watch;create;update;patch;delete
 //+kubebuilder:rbac:groups=db.atlasgo.io,resources=atlasmigrations/status,verbs=get;update;patch
 //+kubebuilder:rbac:groups=db.atlasgo.io,resources=atlasmigrations/finalizers,verbs=update
 
 // Reconcile is part of the main kubernetes reconciliation loop which aims to
 // move the current state of the cluster closer to the desired state.
 func (r *AtlasMigrationReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
-	_ = log.FromContext(ctx)
+	log := log.FromContext(ctx)
+
+	var (
+		am  dbv1alpha1.AtlasMigration
+		err error
+	)
+
+	// Get AtlasMigration resource
+	if err = r.Get(ctx, req.NamespacedName, &am); err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
 
-	// TODO(user): your logic here
+	// At the end of reconcile, update the status of the resource base on the error
+	defer func() {
+		if err != nil {
+			log.Error(err, "failed to reconcile")
+		}
 
+		clientErr := r.updateResourceStatus(ctx, am, err)
+		if clientErr != nil {
+			log.Error(clientErr, "failed to update resource status")
+		}
+	}()
+
+	// Only the 'latest' version is supported
+	if am.Spec.Version != "latest" {
+		err = fmt.Errorf("unsupported version: %s", am.Spec.Version)
+		return ctrl.Result{}, nil
+	}
+
+	// Reconcile given resource
+	am.Status, err = r.reconcile(ctx, am)
+	if err != nil {
+		// Verify permanent errors to avoid infinite reconciliation loop
+		if isPernamentError(err) {
```

Comment:
any reason not to stay consistent with the atlasschema controller? mark errors as transient at their source (the code creating the error has the most context)

Reply:
If you want to emphasize the error most relevant to the context, I suggest creating an error specific to the use case instead of adding additional logic to each function. It will easier when we only find one place that was marked those error as transient

Reply:
We can discuss this further but I prefer to remain consistent with what we have for the atlas schema controller.

Reply:
noted